src/transformers/generation/logits_process.py

```diff
@@ -1102,3 +1103,56 @@ def __call__(self, input_ids, scores):
         cond_logits, uncond_logits = scores.split(unguided_bsz, dim=0)
         scores = uncond_logits + (cond_logits - uncond_logits) * self.guidance_scale
         return scores
+
+
+class UnbatchedClassifierFreeGuidanceLogitsProcessor(LogitsProcessor):
+    r"""Logits processor for Classifier-Free Guidance (CFG). The processors
+    computes a weighted average across scores from prompt conditional and prompt unconditional (or negative) logits,
+    parameterized by the `guidance_scale`. The unconditional scores are computed internally by prompting `model` with
+    the `uncond` branch. Finally, according to CFG Rescale, the reweighted logits are interpolated back with weight
+    `rescale_factor` the conditional ones to smooth the effect and increase output quality.
+
+    See [the paper](https://arxiv.org/abs/2306.17806) for more information.
+
+    Args:
+        guidance_scale (float):
+            The guidance scale for classifier free guidance (CFG). CFG is enabled by setting `guidance_scale > 1`.
+            Higher guidance scale encourages the model to generate samples that are more closely linked to the input
+            prompt, usually at the expense of poorer quality.
+        uncond (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
+            Indices of input sequence tokens in the vocabulary for the unconditional branch.
+        model:
+            The LM computing the unconditional scores. Supposedly the same as the one computing the conditional scores.
+            Both models must use the same tokenizer.
+        smooth_factor (float):
+            The interpolation weight for CFG Rescale. 1 means no rescaling, 0 reduces to the conditional scores without
+            CFG. Turn it lower if the output degenerates.
+    """
+
+    def __init__(self, guidance_scale, uncond, model, rescale_factor=0.7):
+        self.guidance_scale = guidance_scale
+        self.uncond = uncond
+        self.model = model
+        self.out = None
+        self.rescale_factor = rescale_factor
+
+    def __call__(self, input_ids, scores):
+        scores = F.log_softmax(scores, dim=-1)
+        if self.guidance_scale == 1:
+            return scores
+
+        if self.out is None:
+            self.out = self.model(self.uncond, attention_mask=torch.ones_like(self.uncond[:, -1:]), use_cache=True)
```

Comment:
Is it safe to assume the unconditional ids won't use an attention mask?

E.g. if we had a batch size of two, and the following un-conditional prompts:
```
A short prompt
A much much longer one
```
we would need to mask out the padded token ids in the first prompt no? Otherwise the padding is attended to when we run the model forwards

Reply:
Also, if the user doesn't pass unconditional ids (i.e. `self.uncond` is None), this means that they haven't passed a negative prompt. In this case, shouldn't we set the unconditional ids to an unconditional input? i.e. an input of zeros?

Reply:
I'm not comfortable with the attention_mask. I've started NLP (and huggingface) very recently. I will happily take some help on that matter.

Reply:
you don't have enough information to make an attention mask yourself. you should ask the user to provide an uncond attention mask.

if the user submitted two prompts of differing lengths, and the shorter prompt has to be padded to enable them both to be of the same length (so they can reside in the same batch):

```
[[1, 123, 456, 789, 2],
 [1, 123, 2, 3, 3]]
```

the user should also provide an attention mask to say which tokens should be attended to versus which are just padding tokens:

```
[[1, 1, 1, 1, 1],
 [1, 1, 1, 0, 0]]
```

there's nothing you can do on your side to generate this (well, if you knew the tokenizer's conventions you could generate the mask with `self.uncond != PAD_TOKEN`, but that shouldn't be your responsibility.

Reply:
Agree with @Birch-san! The way I see it, we're going to have to get the attention mask for the negative prompt from the `tokenizer`:

```python
inputs = tokenizer(["short prompt", "much longer prompt"], return_tensors="pt")
input_ids = inputs.input_ids
attention_mask = inputs.attention_mask

negative_inputs = tokenizer(["negative prompt", "much longer negative prompt"], return_tensors="pt")
unconditional_ids = negative_inputs.input_ids
unconditional_attention_mask = negative_inputs.attention_mask

# pass both sets of ids and attention mask to the model
model.generate(input_ids, attention_mask=attention_mask, unconditional_ids=unconditional_ids, unconditional_attention_mask=unconditional_attention_mask)
```

(it shouldn't matter that the prompt and negative prompt have different max lengths, since we do `logits[:, -1]` to get the last unconditional logits anyway).

Does that help @Vermeille?

Reply:
Yes it does.